From 72eac48bb9e010a8049da397f61b1fea89099296 Mon Sep 17 00:00:00 2001
From: Jian Zhang <zhangjian210@huawei.com>
Date: Thu, 1 Sep 2022 12:21:46 +0000
Subject: [PATCH] [Huawei] smmu/svsp: alloc pasid when using svm_svsp_bind_core

Offering: HULK
hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I7SZC2

----------------------------------

When using svm_svsp_bind_core, the expected pasid is the pasid of the
origin mm with its higest bit set, so we need to check and get the expected
pasid instead of allocating a new one.

Signed-off-by: Jian Zhang <zhangjian210@huawei.com>
Signed-off-by: Yuan Can <yuancan@huawei.com>
---
 .../iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c   |  4 +++
 drivers/iommu/iommu-sva-lib.c                 | 29 +++++++++++++++++++
 2 files changed, 33 insertions(+)

diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c
index dfc748773699..91ba77a42e7c 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c
@@ -340,7 +340,11 @@ __arm_smmu_sva_bind(struct device *dev, struct mm_struct *mm)
 		return ERR_PTR(-ENOMEM);
 
 	/* Allocate a PASID for this mm if necessary */
+#ifdef CONFIG_ASCEND_SVSP
+	ret = iommu_sva_alloc_pasid(mm, 1, ((1U << master->ssid_bits) - 1) >> 1);
+#else
 	ret = iommu_sva_alloc_pasid(mm, 1, (1U << master->ssid_bits) - 1);
+#endif
 	if (ret)
 		goto err_free_bond;
 
diff --git a/drivers/iommu/iommu-sva-lib.c b/drivers/iommu/iommu-sva-lib.c
index 106506143896..e42ace19af3b 100644
--- a/drivers/iommu/iommu-sva-lib.c
+++ b/drivers/iommu/iommu-sva-lib.c
@@ -10,6 +10,25 @@
 static DEFINE_MUTEX(iommu_sva_lock);
 static DECLARE_IOASID_SET(iommu_sva_pasid);
 
+#ifdef CONFIG_ASCEND_SVSP
+static int iommu_sva_alloc_svsp_pasid(struct mm_struct *mm)
+{
+	void *private;
+	ioasid_t pasid;
+	private = ioasid_find(&iommu_sva_pasid, mm->pasid, NULL);
+	/* reserve the expected SVSP pasid if not found, otherwise do nothing */
+	if (IS_ERR_OR_NULL(private)) {
+		pasid = ioasid_alloc(&iommu_sva_pasid, mm->pasid, mm->pasid, mm);
+		if (pasid == INVALID_IOASID) {
+			pr_err("Failed to reserve SVSP pasid(%d)\n", mm->pasid);
+			return -ENOMEM;
+		}
+	}
+
+	return 0;
+}
+#endif
+
 /**
  * iommu_sva_alloc_pasid - Allocate a PASID for the mm
  * @mm: the mm
@@ -34,6 +53,16 @@ int iommu_sva_alloc_pasid(struct mm_struct *mm, ioasid_t min, ioasid_t max)
 	mutex_lock(&iommu_sva_lock);
 	/* Is a PASID already associated with this mm? */
 	if (pasid_valid(mm->pasid)) {
+#ifdef CONFIG_ASCEND_SVSP
+		/* Pasid which belongs to SVSP has the value of original mm's
+		 * pasid with the highest bit set. When SVSP feature is enabled,
+		 * max is the max value of ssid_bits-1 bits can hold, thus the
+		 * SVSP pasid can exceed max */
+		if (mm->pasid > max) {
+			ret = iommu_sva_alloc_svsp_pasid(mm);
+			goto out;
+		}
+#endif
 		if (mm->pasid < min || mm->pasid >= max)
 			ret = -EOVERFLOW;
 		goto out;
-- 
2.17.1

